src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,358 @@
+    options?: O
+  ): AddListenerAction<ReturnType<C>, S, D, O>
+
+  <
+    S,
+    D extends Dispatch,
+    O extends ActionListenerOptions<AnyAction, S, D, When>
+  >(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, O>,
+    options?: O
+  ): AddListenerAction<AnyAction, S, D, O>
+}
+
+interface RemoveListenerAction<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>
+> {
+  type: 'actionListenerMiddleware/remove'
+  payload: {
+    type: string
+    listener: ActionListener<A, S, D, any>
+  }
+}
+
+/**
+ * @alpha
+ */
+export const removeListenerAction = createAction(
+  'actionListenerMiddleware/remove',
+  function prepare(
+    typeOrActionCreator: string | TypedActionCreator<string>,
+    listener: ActionListener<any, any, any, any>
+  ) {
+    const type =
+      typeof typeOrActionCreator === 'string'
+        ? typeOrActionCreator
+        : (typeOrActionCreator as TypedActionCreator<string>).type
+
+    return {
+      payload: {
+        type,
+        listener
+      }
+    }
+  }
+) as BaseActionCreator<
+  { type: string; listener: ActionListener<any, any, any, any> },
+  'actionListenerMiddleware/remove'
+> & {
+  <C extends TypedActionCreator<any>, S, D extends Dispatch>(
+    actionCreator: C,
+    listener: ActionListener<ReturnType<C>, S, D, any>
+  ): RemoveListenerAction<ReturnType<C>, S, D>
+
+  <S, D extends Dispatch>(
+    type: string,
+    listener: ActionListener<AnyAction, S, D, any>
+  ): RemoveListenerAction<AnyAction, S, D>
+}
+
+/**
+ * @alpha
+ */
+export function createActionListenerMiddleware<
+  S,
+  D extends Dispatch<AnyAction> = Dispatch
+>() {
+  type ListenerEntry = ActionListenerOptions<any, S, D, When> & {
+    listener: ActionListener<any, S, D, any>
+  }
+
+  const listenerMap: Record<string, Set<ListenerEntry> | undefined> = {}
+  const middleware: Middleware<
+    {
+      (action: Action<'actionListenerMiddleware/add'>): Unsubscribe
+    },
+    S,
+    D
+  > = api => next => action => {
+    if (addListenerAction.match(action)) {
+      const unsubscribe = addListener(
+        action.payload.type,
+        action.payload.listener,
+        action.payload.options
+      )
+
+      return unsubscribe
+    }
+    if (removeListenerAction.match(action)) {
+      removeListener(action.payload.type, action.payload.listener)
+
+      return
+    }
+
+    const listeners = listenerMap[action.type]
+    if (listeners) {
+      /* before */
+      for (const entry of listeners) {
+        if (entry.when == 'after') {
+          continue
+        }
+
+        if (!entry.condition || entry.condition(action, api.getState)) {
+          if (entry.once) {
+            listeners.delete(entry)
+          }
+
+          let stoppedPropagation = false
+          entry.listener(action, {
+            ...api,
+            stopPropagation() {
+              stoppedPropagation = true
+            }
+          })
+          if (stoppedPropagation) {
```

Comment:
The current `stopPropagation` idea is troublesome for me.  I can sort of see the intent of wanting to stop other listeners from getting a shot at the action, but I don't think we should ever be preventing the action from going to the next middleware or to the store.  Even with the other listeners I'm struggling the think of a good reason why I'd want need to prevent them getting the action either.  In most cases, I think it will introduce complexity or subtle bugs that are difficult to debug ("Why isn't my action in the dev-tools? _3 weeks later..._ Dammit, the condition on this unrelated listener is too permissive and is smothering it!!!").

To me, it sort of goes beyond its charter as a "listener".  I think if someone wants to have that much control of what actions are able to propagate and which ones can't, then they have the option of using sagas or observables or even writing a custom middleware.

Reply:
The idea of this middleware is to give people an alternative to thunks, as dispatching functions and the usage of higher-order functions is counter-intuitive for many people - they's rather dispatch an action, and react to that. So it should feature-wise be at least en par with thunks while offering another approach.

Now, thunks can take non-serializable parameters, as the thunk itself never reaches the reducer. Likewise, I personally would have "trigger actions" with non-serializable payloads that would never have any effect in any reducer - I'd only use those to trigger my middleware. Did the same thing with sagas.
If those actions would land in the reducer, they would cause problems in the devtools and trigger the nonSerializableActionMiddleware. So there needs to be some way of stopping those actions. Otherwise, listeners would need to start modifying in-transit actions, which would arguably be much worse.

On the flipside, I don't really see the problem you are hinting at - using `stopPropagation` is a *very* deliberate act and nothing that would happen by accident.

On the point of the "listener" name and stopping propagation: Web developers have had "listeners" since the early  days of the DOM. Those "listeners" could always prevent further event propagation. So at least for me, it's a concept that is really entwined with listeners.

Reply:
Ok, I see where you are coming from with this about the serialisable parameters in a trigger action.  I think my preferred usage would be to construct the listener with the non-serialisable aspects and keep the trigger action plain.  This adds an element of tracability, predictablilty and ultimately debuggability to the whole arrangement that, in my experience, is worth the dance around the "actions must be plain" restrictions (especially when working on large teams).

I also feel like this is the first side effect middleware I've seen that lets you swallow an action that isn't directly related to the middleware itself (e.g. the thunk itself, `addListener`/`removeListener`).  I'm not saying that it has to be like everyone else, but I have to wonder why the others didn't need this.

> On the flipside, I don't really see the problem you are hinting at

The problem is that it's a very deliberate action that can be performed on a listener that just happens to be first in the listening order that could prevent a completely unrelated listener (or reducer) from doing it's job.  Here's a scenario (it's a bit contrived, but not _that_ far off something we do where I work, so go with me on this):

We're building a really cool app and one of our devs sees this PR and gets an awesome idea.  What if we could use a listener to listen for a `ROUTE` action to move between routes using a `path` value on the action.  We agree that this could work, so we implement it.

Next someone implements a second listener on `ROUTE` that pings off an analytics event so we can track how our users are navigating through through the app.

A new feature request has come in to make one of the routes dynamic.  Luckily, all the details needed are stored in state, so the `ROUTE` action is updated to pass a function to derive the `path` from state and the middleware gets updated to support this.  By oh no, now we get a warning about the action not being serialisable so we google a bit, find the docs and discover `stopPropagation`, we add it at the end of the listener and away we go.

A few days later we notice a sudden drop in pages getting hit...  You can see where I'm heading with this.

There is very little to help the developer track this down as they can see the action getting dispatched, but their listener is never hit, the dev-tools are not recording it, but the route is definitely changing.

I guess the moral of the story is that is you ship a feature, someone will try to use it, and sometimes that usage will be implemented poorly (there's many things wrong with the above scenario, and many ways it could be worse) and the more someones in the mix, the more chance of something going wrong.  All we can hope for as library authors is when someone gets themselves in a pinch, they find some clue to pull themselves out of it.

> Web developers have had "listeners" since the early days of the DOM. Those "listeners" could always prevent further event propagation. So at least for me, it's a concept that is really entwined with listeners.

This might be where the friction comes from.  I joined the web dev world in the late Angular days and quickly moved to React/Redux, so for me a lot of "early days of the DOM" stuff has either never been used or never been needed, so the idea of preventing propagation on an event bus feels off to me.

At the end of the day, I don't have to like every part of the proposal, so feel free to dismiss this concern.  If I have to choose one method of preventing propagation, the `api. stopPropagation()` is preferred over the other suggestions I read previously.